packages/salesforcedx-vscode-core/test/vscode-integration/commands/auth/forceAuthWebLogin.test.ts

```diff
@@ -243,42 +238,37 @@ describe('Force Auth Web Login is based on environment variables', () => {
     afterEach(() => {
       delete process.env.SFDX_CONTAINER_MODE;
     });
-    it('Should expose the output channel when in container mode', () => {
-      const notContainerMode = new TestForceAuthWebLoginExecutor();
-      expect(notContainerMode.getShowChannelOutput()).to.be.false;
-      process.env.SFDX_CONTAINER_MODE = 'true';
-      const containerMode = new TestForceAuthWebLoginExecutor();
-      expect(containerMode.getShowChannelOutput()).to.be.true;
-    });
-    it('Should use force:auth:web:login when container mode is not defined', () => {
-      const authWebLogin = new ForceAuthWebLoginExecutor();
-      const authWebLoginCommand = authWebLogin.build(
-        ({} as unknown) as AuthParams
-      );
-      expect(authWebLoginCommand.toCommand()).to.equal(
-        'sfdx force:auth:web:login --setalias  --instanceurl  --setdefaultusername'
-      );
+
+    it('Should use ForceAuthWebLoginExecutor when container mode is not defined', () => {
+      expect(createAuthWebLoginExecutor() instanceof ForceAuthWebLoginExecutor)
+        .to.be.true;
     });
 
-    it('Should use force:auth:web:login when container mode is empty', () => {
+    it('Should use ForceAuthWebLoginExecutor when container mode is empty', () => {
       process.env.SFDX_CONTAINER_MODE = '';
-      const authWebLogin = new ForceAuthWebLoginExecutor();
-      const authWebLoginCommand = authWebLogin.build(
-        ({} as unknown) as AuthParams
-      );
-      expect(authWebLoginCommand.toCommand()).to.equal(
-        'sfdx force:auth:web:login --setalias  --instanceurl  --setdefaultusername'
-      );
+      expect(createAuthWebLoginExecutor() instanceof ForceAuthWebLoginExecutor)
+        .to.be.true;
     });
 
-    it('Should use force:auth:device:login when container mode is defined', () => {
+    it('Should use ForceAuthWebLoginContainerExecutor when container mode is defined', () => {
       process.env.SFDX_CONTAINER_MODE = 'true';
-      const authWebLogin = new ForceAuthWebLoginExecutor();
-      const authWebLoginCommand = authWebLogin.build(
-        ({} as unknown) as AuthParams
-      );
+      expect(
+        createAuthWebLoginExecutor() instanceof
+          ForceAuthWebLoginContainerExecutor
+      ).to.be.true;
+    });
+
+    it('should build the force:auth:device:login command', () => {
+      const authWebLogin = new ForceAuthWebLoginContainerExecutor();
+      const authWebLoginCommand = authWebLogin.build({
+        alias: TEST_ALIAS,
+        loginUrl: TEST_URL
+      });
       expect(authWebLoginCommand.toCommand()).to.equal(
-        'sfdx force:auth:device:login --setalias  --instanceurl  --setdefaultusername'
+        `sfdx force:auth:device:login --setalias ${TEST_ALIAS} --instanceurl ${TEST_URL} --setdefaultusername --json --loglevel fatal`
+      );
+      expect(authWebLoginCommand.description).to.equal(
+        nls.localize('force_auth_web_login_authorize_org_text')
       );
     });
```

Comment:
Do we test the execute method and related here? 
potential test cases: 
1. Was openExternal called and did it use the right url based on stdout from the command.
2. The case where the `execution.stdoutSubject.subscribe` is called first with invalid data then a second time with valid data. 
3. The case where the `execution.stdoutSubject.subscribe` is called with unexpected or error data.

Reply:
I was thinking about these tests cases but ended up just going with what the current pattern was in the auth files. We would have to mock the stdOut etc, are there existing examples of that in place I can use?

Honestly, I think I was just lazy and hoping what we have would suffice, sorry 😬 
I'm fine adding these test cases.

Reply:
ha no worries.  This is why code reviews exist :) If you search for `sandbox.stub` or just `stub(` in *.test.* files you'll find a lot of example (probably could get more specific and find exact stubs of the values you'll need to mock in the test file.  Only thing to watch out for is there are some tests that use proxyquire to attempt to be more pure unit tests and those can be more complex to setup.  Feel free to hit me up if you run into any trouble getting stuff mocked.

Reply:
@gbockus-sf I added tests for the cases listed above, please see [chore: refactor for testing](https://github.com/forcedotcom/salesforcedx-vscode/pull/4077/commits/21975b6055a1d0e4586056cb466715fc037fcc6e)

I ended up going a different direction as I couldn't get any of the approaches, based on existing patterns, we discussed to work, so I refactored the code a bit and made it easier to tests just the parts we want.